integration/container/attach_test.go

```diff
@@ -0,0 +1,206 @@
+// +build !windows
+
+package container
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/integration/internal/container"
+	"gotest.tools/assert"
+	is "gotest.tools/assert/cmp"
+	"gotest.tools/poll"
+	"gotest.tools/skip"
+)
+
+// #9860 Make sure attach ends when container ends (with no errors)
+func TestAttachClosedOnContainerStop(t *testing.T) {
+	skip.If(t, testEnv.IsRemoteDaemon, "test requires daemon on the same host")
+
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	cID := container.Run(t, ctx, client, container.WithTty(true),
+		container.WithCmd("/bin/sh", "-c", `trap 'exit 0' SIGTERM; while true; do sleep 1; done`),
+		func(c *container.TestContainerConfig) {
+			c.Config.OpenStdin = true
+		},
+	)
+	defer client.ContainerRemove(context.Background(), cID, types.ContainerRemoveOptions{Force: true})
```

Comment:
Isn't this already taken care of by `defer setupTest(t)()` ? https://github.com/moby/moby/blob/0bb7d42b03bfb125cd50ab50dabe99726e66ab71/integration/container/main_test.go#L30-L33

Reply:
Yes, `setupTest()` takes care of the containers. I have removed all the `defer client.ContainerRemove` calls.